### PR TITLE
chore: [sc-118958] Use Recurly Webhook Type in Smurf

### PR DESCRIPTION
**Pull request by darycabrera**
Story details: https://app.shortcut.com/itprotv/story/118958

---

- [ ] I manually tested the code locally to make sure that it works.
- [ ] If there is documentation, I made sure it's accurate and up to date.
- [ ] If possible, I wrote automated tests for the new behavior.

### CODE REVIEW

File: library/Recurly/V3/Webhook/Webhook.hs [hunk 1]
```diff
@@ -76,7 +80,7 @@ data Notification
   | CreditPaymentVoided Account.Code CreditPayment.Uuid
   | LegacyDunningEventNew Account.Code Invoice.Id Subscription.Uuid Transaction.Id
   | DunningEventNew Account.Code Invoice.Id Subscription.Uuid
-  deriving (Eq, Show)
+  deriving (Eq, Show, Generic, FromJSON, ToJSON, Data.Data)
```
Comment (darycabrera) [inline library/Recurly/V3/Webhook/Webhook.hs]: I was surprised you could derive JSON instances for varying non-unary (I don't know the proper term) sumtypes 🤯

File: library/Recurly/V3/Webhook/Webhook.hs [hunk 2]
```diff
@@ -134,6 +138,58 @@ documentToNotification document = Foldable.asum $ fmap
   , getDunningInfo DunningEventNew "new_dunning_event_notification"
   ]
 
+extractAccountCode :: Notification -> Account.Code
+extractAccountCode notification = case notification of
```
Comment (darycabrera) [inline library/Recurly/V3/Webhook/Webhook.hs:140]: I really don't like this but I could not figure out another way for the design as it stands

File: README.md [hunk 3]
```diff
@@ -1,3 +1,45 @@
 # recurly
 
 haskell partial recurly api
+
+### How to develop recurly alongside Smurf locally
```
Comment (saramuse) [inline README.md]: LOVE this addition! Future us will be thanking you for sure 🤩

File: library/Recurly/V3/Webhook/Webhook.hs (hunk 1 above)
Comment (saramuse) [reply library/Recurly/V3/Webhook/Webhook.hs]: Is this a result of the derived `Data.Data`?

**Review by saramuse — COMMENTED** [review]
Still groking this along with https://github.com/EdutainmentLIVE/smurf/pull/3926 but nothing looks too suspicious to me currently, I am curious what the additions of the `Data.Data` deriving and the two language exceptions `DeriveDataTypeable` + `DeriveAnyClass` are doing to benefit this, as well as any possible pitfalls from these additions.

File: library/Recurly/V3/Webhook/Webhook.hs (hunk 2 above)
Comment (saramuse) [reply library/Recurly/V3/Webhook/Webhook.hs:140]: It's a bit beefy but I don't have a suggestion currently, for our purposes I think this should be fine.

File: library/Recurly/V3/Webhook/Webhook.hs (hunk 1 above)
Comment (tfausak) [reply library/Recurly/V3/Webhook/Webhook.hs]: It's from the `Generic` instance. I'm surprised to see `Data` there, actually. What's it used for?

File: library/Recurly/V3/Webhook/Webhook.hs (hunk 2 above)
Comment (tfausak) [reply library/Recurly/V3/Webhook/Webhook.hs:140]: There's no other way to write this. Looks reasonable to me, except that you don't need parens around the pattern matches and you could use a shorter variable name (like `x`) to make it visually simpler.

File: library/Recurly/V3/Webhook/Webhook.hs (hunk 1 above)
Comment (darycabrera) [reply library/Recurly/V3/Webhook/Webhook.hs]: Its used in soley for the `Data.toConstr` function in  smurf's `WebhookProcessor.hs`. I left a comment about that in there. I'll probably end up removing it

File: README.md (hunk 3 above)
Comment (saramuse) [reply README.md]: Just a reminder for all @EdutainmentLIVE/engineers - this is a public facing repository, so we should probably change it to private before merging this change.

Comment (saramuse) [reply README.md]: https://app.shortcut.com/itprotv/story/122138/move-recurly-into-smurf

**Review by saramuse — APPROVED** [review]

**Pull request merged**